File: app/models/niche_config.py
"""NicheConfig model — stores niche configuration per user/brand."""

from datetime import datetime
from sqlalchemy import Column, String, Text, DateTime, UniqueConstraint, ForeignKey
from sqlalchemy.dialects.postgresql import JSONB
from app.models.base import Base
import uuid


class NicheConfig(Base):
    __tablename__ = "niche_config"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(100), nullable=False)

    # NULL brand_id = global config; non-NULL = per-brand override
    brand_id = Column(String(50), ForeignKey("brands.id", ondelete="CASCADE"), nullable=True)

    # Core Identity — all empty by default, user fills in everything
    niche_name = Column(String(100), nullable=False, default="")
    niche_description = Column(Text, default="")
    content_brief = Column(Text, default="")
    target_audience = Column(String(255), default="")
    audience_description = Column(Text, default="")
    content_tone = Column(JSONB, default=[])
    tone_avoid = Column(JSONB, default=[])

    # Topic Configuration — empty, user defines their own niche topics
    topic_categories = Column(JSONB, default=[])
    topic_keywords = Column(JSONB, default=[])
    topic_avoid = Column(JSONB, default=[])

    # Content Philosophy
    content_philosophy = Column(Text, default="")
    hook_themes = Column(JSONB, default=[])

    # User Examples (few-shot prompting)
    reel_examples = Column(JSONB, default=[])
    post_examples = Column(JSONB, default=[])

    # Visual Configuration
    image_style_description = Column(Text, default="")
    image_palette_keywords = Column(JSONB, default=[])

    # Brand Personality
    brand_personality = Column(Text, nullable=True)
    brand_focus_areas = Column(JSONB, default=[])
    parent_brand_name = Column(String(100), default="")

    # CTA Configuration — empty, user adds their own CTAs
    cta_options = Column(JSONB, default=[])
    hashtags = Column(JSONB, default=[])

    # Discovery Configuration — empty, user defines competitor accounts and hashtags
    competitor_accounts = Column(JSONB, default=[])
    discovery_hashtags = Column(JSONB, default=[])

    # Citation / source style — determines carousel post source format
    # Values: "academic_doi" | "financial_data" | "case_study" | "expert_quote" | "none"
    citation_style = Column(String, default="none")

    # Source databases/organizations for citations (overrides built-in defaults when set)
    citation_source_types = Column(JSONB, default=[])

    # Visual composition style — replaces REEL_BASE_STYLE per niche
    image_composition_style = Column(Text, default="")

    # YouTube title examples — good titles that show format for this niche
    yt_title_examples = Column(JSONB, default=[])

    # YouTube title bad examples — titles to avoid for this niche
    yt_title_bad_examples = Column(JSONB, default=[])

    # Carousel CTA topic word — the niche-relevant word used in slide 4 CTA
    carousel_cta_topic = Column(String, default="")

    # Caption sections — empty until configured
    follow_section_text = Column(Text, default="")
    save_section_text = Column(Text, default="")
    disclaimer_text = Column(Text, default="")

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    __table_args__ = (
        UniqueConstraint("user_id", "brand_id", name="uq_niche_config_user_brand"),
    )

File: src/features/brands/types/niche-config.ts
export interface ReelExample {
  title: string
  content_lines: string[]
}

export interface PostExample {
  title: string
  slides: string[]
  doi?: string
  /** Max slides for this example (3 or 4). Not persisted — UI-only. */
  _maxSlides?: number
}

export interface CtaOption {
  text: string
  weight: number  // percentage 0-100, all weights must sum to 100
}

export interface NicheConfig {
  id?: string | null
  brand_id?: string | null

  // Core Identity
  niche_name: string
  niche_description: string
  content_brief: string
  target_audience: string
  audience_description: string
  content_tone: string[]
  tone_avoid: string[]

  // Topics
  topic_categories: string[]
  topic_keywords: string[]
  topic_avoid: string[]

  // Content Philosophy
  content_philosophy: string
  hook_themes: string[]

  // Examples
  reel_examples: ReelExample[]
  post_examples: PostExample[]

  // Visual
  image_style_description: string
  image_palette_keywords: string[]
  image_composition_style: string

  // Brand Identity
  brand_personality: string | null
  brand_focus_areas: string[]
  parent_brand_name: string

  // CTAs
  cta_options: CtaOption[]
  hashtags: string[]
  carousel_cta_topic: string
  follow_section_text: string
  save_section_text: string
  disclaimer_text: string

  // Citation & YouTube
  citation_style: string
  citation_source_types: string[]
  yt_title_examples: string[]
  yt_title_bad_examples: string[]
}

export type ConfigStrength = 'basic' | 'good' | 'excellent'

export function getConfigStrength(config: NicheConfig): ConfigStrength {
  let score = 0
  const maxScore = 12

  // Core identity (3 points)
  if (config.niche_name) score++
  if (config.content_brief && config.content_brief.length > 50) score++
  if (config.content_brief && config.content_brief.length > 200) score++

  // Examples (3 points)
  if (config.reel_examples.length >= 3) score++
  if (config.reel_examples.length >= 10) score++
  if (config.post_examples.length >= 1) score++

  // CTAs (1 point)
  if (config.cta_options.length > 0 && config.cta_options.some(c => c.text.trim())) score++

  // Visual & content style (2 points)
  if (config.image_composition_style && config.image_composition_style.trim()) score++
  if (config.carousel_cta_topic && config.carousel_cta_topic.trim()) score++

  // YouTube titles (1 point)
  if ((config.yt_title_examples || []).length >= 2) score++

  // Citation (1 point)
  if (config.citation_style && config.citation_style !== 'none') score++

  // Brand name (1 point)
  if (config.parent_brand_name && config.parent_brand_name.trim()) score++

  const pct = score / maxScore
  if (pct < 0.4) return 'basic'
  if (pct >= 0.75) return 'excellent'
  return 'good'
}

export function getStrengthLabel(strength: ConfigStrength): string {
  switch (strength) {
    case 'basic': return 'Add more details and examples to improve content quality'
    case 'good': return 'Good start! Adding examples will take it to the next level.'
    case 'excellent': return 'Your AI knows exactly what to generate.'
  }
}

export function getStrengthColor(strength: ConfigStrength): string {
  switch (strength) {
    case 'basic': return 'text-red-500'
    case 'good': return 'text-yellow-500'
    case 'excellent': return 'text-green-500'
  }
}

export function getStrengthBarColor(strength: ConfigStrength): string {
  switch (strength) {
    case 'basic': return 'bg-red-500'
    case 'good': return 'bg-yellow-500'
    case 'excellent': return 'bg-green-500'
  }
}

export function getStrengthPercent(strength: ConfigStrength): number {
  switch (strength) {
    case 'basic': return 25
    case 'good': return 65
    case 'excellent': return 100
  }
}
